ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,932 @@
+    const tradeFee = this.calculateTradingFee(sourceAmount);
+
+    const sourceVaultWithdrawableAmount = await swapSourceVault.getWithdrawableAmount();
+    // Get vault lp minted when deposit to the vault
+    const sourceVaultLp = getUnmintAmount(
+      sourceAmount.sub(adminFee),
+      sourceVaultWithdrawableAmount,
+      swapSourceVaultLpSupply,
+    );
+
+    const actualSourceAmount = getAmountByShare(sourceVaultLp, sourceVaultWithdrawableAmount, swapSourceVaultLpSupply);
+
+    let sourceAmountWithFee = actualSourceAmount.sub(tradeFee);
+
+    const destinationAmount = this.swapCurve.computeOutAmount(
+      sourceAmountWithFee,
+      swapSourceAmount,
+      swapDestinationAmount,
+      tradeDirection,
+    );
+
+    const destinationVaultWithdrawableAmount = await swapDestinationVault.getWithdrawableAmount();
+    // Get vault lp to burn when withdraw from the vault
+    const destinationVaultLp = getUnmintAmount(
+      destinationAmount,
+      destinationVaultWithdrawableAmount,
+      swapDestinationVaultLpSupply,
+    );
+
+    let actualDestinationAmount = getAmountByShare(
+      destinationVaultLp,
+      destinationVaultWithdrawableAmount,
+      swapDestinationVaultLpSupply,
+    );
+
+    return getMinAmountWithSlippage(actualDestinationAmount, slippageRate);
+  }
+
+  /**
+   * `getMaxSwapOutAmount` returns the maximum amount of tokens that can be swapped out of the pool
+   * @param {PublicKey} tokenMint - The mint of the token you want to swap out.
+   * @returns The maximum amount of tokens that can be swapped out of the pool.
+   */
+  public async getMaxSwapOutAmount(tokenMint: PublicKey) {
+    invariant(
+      tokenMint.equals(this.poolState.tokenAMint) || tokenMint.equals(this.poolState.tokenBMint),
+      ERROR.INVALID_MINT,
+    );
+
+    const { tokenAAmount, tokenBAmount } = await this.getPoolInfo();
+    const { vaultAReserve, vaultBReserve } = await this.getSplInfo();
+    const [outTotalAmount, outReserveBalance] = tokenMint.equals(this.poolState.tokenAMint)
+      ? [tokenAAmount, vaultAReserve]
+      : [tokenBAmount, vaultBReserve];
+
+    return outTotalAmount.gt(outReserveBalance) ? outReserveBalance : outTotalAmount;
+  }
+
+  /**
+   * `swap` is a function that takes in a `PublicKey` of the owner, a `PublicKey` of the input token
+   * mint, an `BN` of the input amount of lamports, and an `BN` of the output amount of lamports. It
+   * returns a `Promise<Transaction>` of the swap transaction
+   * @param {PublicKey} owner - The public key of the user who is swapping
+   * @param {PublicKey} inTokenMint - The mint of the token you're swapping from.
+   * @param {BN} inAmountLamport - The amount of the input token you want to swap.
+   * @param {BN} outAmountLamport - The amount of the output token you want to receive.
+   * @returns A transaction object
+   */
+  public async swap(owner: PublicKey, inTokenMint: PublicKey, inAmountLamport: BN, outAmountLamport: BN) {
```

Comment:
swap function no need make it sync, jup integration will write their own swap ix